examples/pytorch/bgnn/BGNN.py

```diff
@@ -0,0 +1,356 @@
+        for metric_name in train_results:
+            metrics[metric_name].append((train_results[metric_name].detach().item(),
+                               val_results[metric_name].detach().item(),
+                               test_results[metric_name].detach().item()
+                               ))
+        return loss
+
+    def update_early_stopping(self, metrics, epoch, best_metric, best_val_epoch, epochs_since_last_best_metric, metric_name,
+                              lower_better=False):
+        train_metric, val_metric, test_metric = metrics[metric_name][-1]
+        if (lower_better and val_metric < best_metric[1]) or (not lower_better and val_metric > best_metric[1]):
+            best_metric = metrics[metric_name][-1]
+            best_val_epoch = epoch
+            epochs_since_last_best_metric = 0
+        else:
+            epochs_since_last_best_metric += 1
+        return best_metric, best_val_epoch, epochs_since_last_best_metric
+
+    def log_epoch(self, pbar, metrics, epoch, loss, epoch_time, logging_epochs, metric_name='loss'):
+        train_metric, val_metric, test_metric = metrics[metric_name][-1]
+        if epoch and epoch % logging_epochs == 0:
+            pbar.set_description(
+                "Epoch {:05d} | Loss {:.3f} | Loss {:.3f}/{:.3f}/{:.3f} | Time {:.4f}".format(epoch, loss,
+                                                                                              train_metric,
+                                                                                              val_metric,
+                                                                                              test_metric,
+                                                                                              epoch_time))
+
+    def fit(self, graph, X, y,
+            train_mask, val_mask, test_mask,
+            cat_features = None,
+            num_epochs=100,
+            patience=10,
+            logging_epochs=1,
+            metric_name='loss',
+            normalize_features=False,
+            replace_na=True,
+            ):
+        '''
+
+        :param graph: dgl graph
+        :param X: numpy matrix of input features. Can be composed of categorical features with missing values.
+        :param y: numpy matrix of target labels. Can be multidimensional targets.
+        :param train_mask: list of integers (rows in X that correspond to train set)
+        :param val_mask: list of integers (rows in X that correspond to valid set)
+        :param test_mask: list of integers (rows in X that correspond to test set)
+        :param cat_features: list of integers (columns in X that correspond to categorical features)
+        :param num_epochs: number of epochs to run
+        :param patience: number of epochs to wait until early stopping
+        :param logging_epochs: log every n epoch
+        :param metric_name: metric to use for early stopping
```

Comment:
I wonder how different early stopping metric performs in practice? Is it critical? Can we remove the option and use the default value `'loss'`?

Reply:
I think we better keep as it is, as we want to look at accuracy and not at the loss, if accuracy if our final goal.